### PR TITLE
Recursive attribute conversion

### PR DESCRIPTION
**Pull request by fabianmarz**
- Added recursive attributes conversion for components.
- Added Twig link function.

### CODE REVIEW

File: src/attributes.js [hunk 1]
```diff
@@ -93,6 +93,21 @@ class Attributes {
         return this;
     };
 
+    // Class names of variables whose names contain 'attributes' are merged
+    // instead of being replaced.
+    // @see TwigAdapter:Twig.Template.prototype.render()
+    // @see TwigAdapter/tags:render()
```
Comment (sun) [inline src/attributes.js]: 1. Should use JSDoc block syntax instead of inline comments.

2. Static helper functions should be declared somewhere else than the middle of all other functions; either at the very beginning or at the very end of the class.

**Review by sun — CHANGES_REQUESTED** [review]

File: src/attributes.js (hunk 1 above)
Comment (fabianmarz) [reply src/attributes.js]: You're right. Since it was called `convertAttributes()` I wanted to stay it close to the other `<x>Attribute()` methods. Since I changed the name while `Attribute` became obsolete it makes more sense to place it somewhere else. Thanks for pointing out.

File: src/attributes.js [hunk 2]
```diff
@@ -131,6 +131,29 @@ class Attributes {
         return this;
     };
 
+    /**
+    * Recursively convert context 'attribute' keys to an instance of this class.
```
Comment (sun) [inline src/attributes.js]: Wrong indentation / JSDoc block formatting.

File: src/attributes.js [hunk 3]
```diff
@@ -131,6 +131,29 @@ class Attributes {
...
+    *
+    * Class names of variables whose names contain 'attributes' are merged
+    * instead of being replaced.
+    *
+    * @see TwigAdapter:Twig.Template.prototype.render()
+    * @see TwigAdapter/tags:render()
+    *
+    * @param mixed context
```
Comment (sun) [inline src/attributes.js]: The passed context is always of type `object`

File: src/functions/index.js [hunk 4]
```diff
@@ -10,6 +11,14 @@ module.exports = function (fractal) {
         },
         url(str, obj) {
             return 'url://' + str + queryString.stringify(obj);
+        },
+        link(text, url = '', attributes = {}) {
```
Comment (sun) [inline src/functions/index.js:15]: Happy to handle this in the same PR, but it should be a separate commit that is not a fixup and instead with proper commit message that explains what this change is about.

**Review by sun — CHANGES_REQUESTED** [review]
Do you want to prepare the two final commits already?

**Review by sun — APPROVED** [review]
Fixed the JSDoc + docs, and upgraded the second instance of Attributes conversion in the render tag code to the new helper function.

**Comment by sun** [comment]
Hm, you did not really account for the URL input values especially concerning `internal:` and `base:` schemes here… https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Url.php/function/Url%3A%3AfromUri/8.2.x

**Comment by fabianmarz** [comment]
Sorry, I don't get your comment. Could you please explain it a bit further?

**Comment by sun** [comment]
Did you read the documentation page?  What exactly is not clear for you?

The Twig `link()` function uses aforementioned Url method.  https://api.drupal.org/comment/61905#comment-61905 shows examples for valid values you can pass to it.  Any other values are invalid and cause an exception.

**Comment by fabianmarz** [comment]
Ah sorry, now I get it. Yes, I did not account such values as they should not be denoted in the component config, but rather coming from the actual Drupal Menu. The helper is functioned just as a simple link function which converts given arguments. There would just be an advantage if we use the Fractal component data as a reusable menu without passing custom items, or do you think it is really necessary to consider "special" values also?

**Comment by sun** [comment]
The default values in the component definitions need to be compatible with Drupal.  Because if the values are not compatible, and Drupal passes no actual / dynamic values, then the site renders a fatal error / uncaught exception, which means an error in code, which must not happen under any circumstances.
